.golangci.yml

```diff
@@ -0,0 +1,199 @@
+      require-explanation: true
+      allow-unused: false
+      require-specific: true
+
+  revive:
+    ignore-generated-header: true
+    severity: warning
+    rules:
+      - name: atomic
+      - name: blank-imports
+      - name: bool-literal-in-expr
+      - name: confusing-naming
+      - name: constant-logical-expr
+      - name: context-as-argument
+      - name: context-keys-type
+      - name: deep-exit
+      - name: defer
+      - name: range-val-in-closure
+      - name: range-val-address
+      - name: dot-imports
+      - name: error-naming
+      - name: error-return
+      - name: error-strings
+      - name: errorf
+      - name: exported
+      - name: identical-branches
+      - name: if-return
+      - name: import-shadowing
+      - name: increment-decrement
+      - name: indent-error-flow
+      - name: indent-error-flow
+      - name: package-comments
+      - name: range
+      - name: receiver-naming
+      - name: redefines-builtin-id
+      - name: superfluous-else
+      - name: struct-tag
+      - name: time-naming
+      - name: unexported-naming
+      - name: unexported-return
+      - name: unnecessary-stmt
+      - name: unreachable-code
+      - name: unused-parameter
+      - name: var-declaration
+      - name: var-naming
+      - name: unconditional-recursion
+      - name: waitgroup-by-value
+
+  staticcheck:
+    go: "1.16"
+
+  unused:
+    go: "1.16"
+
+output:
+  sort-results: true
+
+linters:
+  disable-all: true
+  enable:
+    - asciicheck
+    - bodyclose
+    - cyclop
+    - deadcode
+    - dogsled
+    - dupl
+    - durationcheck
+    - errcheck
+    # errname is only available in golangci-lint v1.42.0+ - wait until v1.43 is available to settle
+    #- errname
+    - errorlint
+    - exhaustive
+    - exportloopref
+    - forcetypeassert
+    - gocognit
+    - goconst
+    - gocritic
+    - godot
+    - gofmt
+    - goheader
+    - goimports
+    - goprintffuncname
+    - gosimple
+    - govet
+    - ifshort
+    - importas
+    - ineffassign
+    - makezero
+    - misspell
+    - nakedret
+    - nestif
+    - nilerr
+    - noctx
+    - nolintlint
+    - predeclared
+    # disabling for the initial iteration of the linting tool
+    #- promlinter
+    - revive
+    - rowserrcheck
+    - sqlclosecheck
+    - staticcheck
+    - structcheck
+    - stylecheck
+    - thelper
+    - tparallel
+    - typecheck
+    - unconvert
+    - unparam
+    - unused
+    - varcheck
+    - wastedassign
+    - whitespace
+
+  # Disabled linters, due to being misaligned with Go practices
+  # - exhaustivestruct
+  # - gochecknoglobals
+  # - gochecknoinits
+  # - goconst
+  # - godox
+  # - goerr113
+  # - gomnd
+  # - lll
+  # - nlreturn
+  # - testpackage
+  # - wsl
+
+  # Disabled linters, due to not being relevant to our code base:
+  # - maligned
+  # - prealloc "For most programs usage of prealloc will be a premature optimization."
+
+  # Disabled linters due to bad error messages or bugs
+  # - gofumpt
```

Comment:
It simply emits a message like `gofumpt was not run on this file`, instead of highlighting the actual lint issue that should be resolved, like `out-of-order imports: try running gofumpt`.

If gofumpt seems like an acceptable requirement, I may add it in a subsequent PR to lint-install. I just wish their error was better.

Reply:
> It simply emits a message like `gofumpt was not run on this file`, instead of highlighting the actual lint issue that should be resolved, like `out-of-order imports: try running gofumpt`.
> 
> If gofumpt seems like an acceptable requirement, I may add it in a subsequent PR to lint-install. I just wish their error was better.
 
`goimports` has similar behavior, kind of. We've worked around it by calling it with `-w` so that it does it's thing and then we check if any files are changed using `git-diff`. This also helps to show the diff too. I think `gofumpt` is meaning to emulate `goimports` so we may never get better behavior.
 
> One open question is what to do with autogenerated Go files that aren't gofumpt'd.

Anything that makes it into git is also `goimports` usually, so might as well do it here too.

Reply:
ok. I'm still going to leave gofumpt behavioral changes out of this PR. 

One additional point of hesitation is that the changes suggested by by `gofumpt` are not seen as idiomatic or necessary for Go. Running it against the Go code base shows that gofumpt would make changes to 1718 of its Go files (22% of the code base). 

That said, I continue to use gofumpt myself, even if it doesn't seem to be universally in use yet.

Reply:
> ok. I'm still going to leave gofumpt behavioral changes out of this PR.

Yep, we should think about gofumpt separately.
  
> 
> One additional point of hesitation is that the changes suggested by by `gofumpt` are not seen as idiomatic or necessary for Go. Running it against the Go code base shows that gofumpt would make changes to 1718 of its Go files (22% of the code base).
> 
> That said, I continue to use gofumpt myself, even if it doesn't seem to be universally in use yet.

I ran gofumpt on some of our tinkerbell code bases iirc and really liked the changes, mostly just as an increase in consistency.

Reply:
I agree as well. gofumpt was the first command I ran on this PR :) 

I'll see what I can do to integrate gofumpt in a way that doesn't result in terrible error messages, and avoids at least some of the auto-generated files by default (protobuf code, for instance). I've opened https://github.com/tinkerbell/lint-install/issues/11